File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/ui/settings/view/SettingsFragment.java
package com.vereshchagin.nikolay.stankinschedule.ui.settings.view;


import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;

import androidx.annotation.IdRes;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import com.vereshchagin.nikolay.stankinschedule.R;

/**
 * Фрагмент с настройками приложения.
 */
public class SettingsFragment extends PreferenceFragmentCompat
        implements Preference.OnPreferenceClickListener {

    private static final String SETTINGS_GENERAL_FRAGMENT = "settings_general_fragment";
    private static final String SETTINGS_SCHEDULE_FRAGMENT = "settings_schedule_fragment";
    private static final String SETTINGS_WIDGET_FRAGMENT = "settings_widget_fragment";
    private static final String SETTINGS_NOTIFICATION_FRAGMENT = "settings_notification_fragment";

    public SettingsFragment() {
        super();
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        setPreferencesFromResource(R.xml.preferences_root, rootKey);

        Preference preferenceGeneral = findPreference(SETTINGS_GENERAL_FRAGMENT);
        if (preferenceGeneral!= null) {
            preferenceGeneral.setOnPreferenceClickListener(this);
        }

        Preference preferenceSchedule = findPreference(SETTINGS_SCHEDULE_FRAGMENT);
        if (preferenceSchedule != null) {
            preferenceSchedule.setOnPreferenceClickListener(this);
        }

        Preference preferenceWidget = findPreference(SETTINGS_WIDGET_FRAGMENT);
        if (preferenceWidget != null) {
            preferenceWidget.setOnPreferenceClickListener(this);
        }

        Preference preferenceNotification = findPreference(SETTINGS_NOTIFICATION_FRAGMENT);
        if (preferenceNotification != null) {
            preferenceNotification.setOnPreferenceClickListener(this);

        }
    }
    @Override
    public boolean onPreferenceClick(Preference preference) {
        if (getActivity() == null) {
            return false;
        }

        // реализация передвижения по настройкам с помощью Navigation graph
        NavController navController = Navigation.findNavController(getActivity(), R.id.nav_host);

        @IdRes
        int destination;

        switch (preference.getKey()) {
            // к основным настройкам
            case SETTINGS_GENERAL_FRAGMENT:
                destination = R.id.toSettingsGeneralFragment;
                break;
            // к настройкам расписания
            case SETTINGS_SCHEDULE_FRAGMENT:
                destination = R.id.toSettingsScheduleFragment;
                break;
            // к настройкам виджетов
            case SETTINGS_WIDGET_FRAGMENT:
                destination = R.id.toSettingsWidgetFragment;
                break;
            // к настройкам уведомлений
            case SETTINGS_NOTIFICATION_FRAGMENT:
                // android 8.0+
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    Intent  intent = new Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS);
                    intent.putExtra(Settings.EXTRA_APP_PACKAGE, requireContext().getPackageName());
                    startActivity(intent);

                    return true;
                }

                destination = R.id.toSettingsNotificationFragment;
                break;
            default:
                return false;
        }

        // TODO("07/02/21 Ошибка о переходе из "Основных" настроек в настройки "расписания"")
        navController.navigate(destination);
        return true;
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/ui/settings/view/SettingsFragment.kt
package com.vereshchagin.nikolay.stankinschedule.ui.settings.view

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import androidx.annotation.IdRes
import androidx.navigation.Navigation
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import com.vereshchagin.nikolay.stankinschedule.R

/**
 * Фрагмент с настройками приложения.
 */
class SettingsFragment : PreferenceFragmentCompat(), Preference.OnPreferenceClickListener {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.preferences_root, rootKey)

        findPreference<Preference>(SETTINGS_GENERAL_FRAGMENT)?.onPreferenceClickListener = this
        findPreference<Preference>(SETTINGS_SCHEDULE_FRAGMENT)?.onPreferenceClickListener = this
        findPreference<Preference>(SETTINGS_WIDGET_FRAGMENT)?.onPreferenceClickListener = this
        findPreference<Preference>(SETTINGS_NOTIFICATION_FRAGMENT)?.onPreferenceClickListener = this

    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        // реализация передвижения по настройкам с помощью Navigation graph
        val navController = Navigation.findNavController(requireActivity(), R.id.nav_host)

        @IdRes
        val destination: Int = when (preference.key) {
            SETTINGS_GENERAL_FRAGMENT -> R.id.toSettingsGeneralFragment
            SETTINGS_SCHEDULE_FRAGMENT -> R.id.toSettingsScheduleFragment
            SETTINGS_WIDGET_FRAGMENT -> R.id.toSettingsWidgetFragment
            SETTINGS_NOTIFICATION_FRAGMENT -> {
                // android 8.0+
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    val intent = Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS)
                    intent.putExtra(Settings.EXTRA_APP_PACKAGE, requireContext().packageName)
                    startActivity(intent)
                    return true
                }
                R.id.toSettingsNotificationFragment
            }
            else -> return false
        }

        // TODO("07/02/21 Ошибка о переходе из "Основных" настроек в настройки "расписания"")
        navController.navigate(destination)
        return true
    }

    companion object {
        private const val SETTINGS_GENERAL_FRAGMENT = "settings_general_fragment"
        private const val SETTINGS_SCHEDULE_FRAGMENT = "settings_schedule_fragment"
        private const val SETTINGS_WIDGET_FRAGMENT = "settings_widget_fragment"
        private const val SETTINGS_NOTIFICATION_FRAGMENT = "settings_notification_fragment"
    }
}